src/require-login.tsx

```diff
@@ -0,0 +1,171 @@
+
+      client.use(oauth2({
+        grantType: 'authorization_code',
+        clientId: props.clientId,
+        tokenEndpoint: props.tokenEndpoint,
+        redirectUri: document.location.origin + '/',
+        code: code,
+        onTokenUpdate: (token: TokenSet) => storeTokens(token),
+      }));
+
+      setAuthState('creds-setup');
+      const history = useHistory();
+      history.push(redirectAfter);
+    }
+
+    const processTokenFromLS = (tokens: TokenSet) => {
+      client.use(oauth2({
+        grantType: undefined,
+        clientId: props.clientId,
+        tokenEndpoint: props.tokenEndpoint,
+        onTokenUpdate: (token: TokenSet) => storeTokens(token),
+        refreshToken: tokens.refreshToken!,
+        accessToken: tokens.accessToken,
+      }));
+      setAuthState('creds-setup');
+
+    };
+
+    const initiateAuthFlow = () => {
+
+      const currentRoot = document.location.origin + '/';
+      const currentPath = document.location.pathname;
+
+      const params = new URLSearchParams([
+        ['response_type', 'code'],
+        ['client_id', props.clientId],
+        ['redirect_uri', currentRoot],
+        ['state', currentPath],
+      ]);
+
+      console.log('Redirecting to OAuth2 authorization endpoint');
+      document.location.href = props.authorizeEndpoint + '?' + params.toString();
+      
+    };
+
+    const testCredentials = () => {
+
+      const testResource = props.testEndpoint instanceof Resource ? props.testEndpoint : client.go(props.testEndpoint);
+      
+      testResource.refresh()
+        .then( () => {
+          console.log('Credentials accepted');
+          setAuthState('authenticated');
+        })
+        .catch( (err:any) => {
+          if (err.httpCode !== 401) {
+            const newErr = new Error('Got error when accessing test endpoint: ' + err.httpCode);
+            console.error(newErr);
+            console.error(err);
+            throw newErr;
+          } else {
+            console.log('Credentials rejected. Attempting a fresh login');
+            initiateAuthFlow();
+          }
+        });
+
+    }
+
+    if (authState === 'no-auth') {
+      // First, we should see if we got here after a redirect back
+      // an OAuth2 authentication process.
+      //
+      // If this is the case, we'll have a `code` parameter in the url.
+      const location = new URL(document.location.href);
+      if (location.search) {
+
+        const searchParams = new URLSearchParams(location.search.substr(1));
+        if (searchParams.has('code')) {
+          console.log('Got authorization code from redirect');
+          const code: string = searchParams.get('code')!;
+          const redirectAfter: string = searchParams.get('state') || '/';
+          processCodeFromUrl(code, redirectAfter);
+          return;
+        }
+      }
+
+      // Do we have credentials in LocalStorage?
+      const ls = window.localStorage.getItem('ketting-auth:v3');
```

Comment:
Is it okay to use localStorage here? I believe from  https://dev.to/rdegges/please-stop-using-local-storage-1i04 suggests that those tokens can be leaked to any thirdparty scripts exectued on the page.